### PR TITLE
Converters: use target datatype

### PR DESCRIPTION
**Pull request by bernt-matthias**
https://github.com/galaxyproject/galaxy/pull/12104#issuecomment-857726453.

The PR: 

- introduces a variable `$__target_datatype__` (see linked comment above) that allows to reduce the number of needed converters
- show use for the biom converters which are unified to a single converter
- add a converter test for the biom converters
 
I would like to stop with the biom converter here and implement the other tests here https://github.com/galaxyproject/galaxy/pull/12104 (hoping to avoid conflicts).

TODO

- [x] biom1<->biom2
  - [x] one could test in the `if-else` statement that only the proper combinations of input and output type are recieved
- [x] document `$__target_datatype` ..  maybe in galaxy.xsd -- or is there some special docs for converters?

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

File: lib/galaxy/datatypes/converters/biom.xml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        <requirement type="package" version="2.1.5">biom-format</requirement>
+    </requirements>
+    <command><![CDATA[
+        biom convert -i '$input' -o '$output'
+        #if $__target_datatype == 'biom2'
+            --to-hdf5
+        #elif $__target_datatype == 'biom1'
+            --to-json
+        #else
+            >&2 echo "BIOM converter recieved unknown target datatype: "$__target_datatype &&
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/converters/biom.xml]:
```suggestion
            >&2 echo "BIOM converter received unknown target datatype: $__target_datatype" &&
```

**Review by nsoranzo — COMMENTED** [review]
I am not sure I like this direction, it seems too much "magic" for such simple tools. Splitting the testing in another file is another downside IMO.

**Comment by mvdbeek** [comment]
Can you make `__target_datatype` a regular select parameter in the tool ? That would make the tool testable with our standard tool testing and would mean that you can use the converter explicitly.

**Comment by bernt-matthias** [comment]
> Can you make __target_datatype a regular select parameter in the tool ? That would make the tool testable with our standard tool testing and would mean that you can use the converter explicitly.

That's an excellent idea. Guess this would also make @nsoranzo happier :)

File: lib/galaxy/datatypes/converters/biom.xml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        <configfile name="ext_config">{"output": {
+  "name": "$input.name converted to $__target_datatype",
+  "ext": "$__target_datatype"
+}}</configfile>
+    </configfiles>
+    <inputs>
+        <param name="input" type="data" format="biom1,biom2" label="Biom file"/>
+        <!-- this hidden parameter is set automatically when the tool is called as
+             converter and set manually when called from the tool test -->
+        <param name="__target_datatype" type="hidden" label="Target data type"/>
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/converters/biom.xml]: Can you make this a normal select parameter? That wait you can skip the else branch (`BIOM converter received unknown target datatype: $__target_datatype ...`) and users can set the parameter in the user interface.

Comment (bernt-matthias) [reply lib/galaxy/datatypes/converters/biom.xml]: Yep. Where in the UI could users select the target datatype? In the dataset properties convert tab?

Comment (mvdbeek) [reply lib/galaxy/datatypes/converters/biom.xml]: The plan is to expose all converters in the user interface, but for now you need to know the url https://usegalaxy.org/root?tool_id=sam_to_bam

Comment (bernt-matthias) [reply lib/galaxy/datatypes/converters/biom.xml]: Thanks for the explanation. In this case I guess the biom converter should be excluded https://github.com/galaxyproject/tools-iuc/blob/master/tools/biom_format/biom_convert.xml :)

File: lib/galaxy/datatypes/converters/biom.xml [hunk 3]
```diff
@@ -0,0 +1,51 @@
+<tool id="CONVERTER_biom" name="Convert Biom datasets" version="2.1.5">
+    <requirements>
+        <requirement type="package" version="2.1.5">biom-format</requirement>
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/converters/biom.xml]: Macro for the requirement version to use also in tool version?

File: lib/galaxy/datatypes/converters/biom.xml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        </param>
+    </inputs>
+    <outputs>
+        <data name="output" format="auto"/>
+    </outputs>
+    <tests>
+        <test>
+            <param name="input" ftype="biom2" value="input_taxonomy.biom2"/>
+            <param name="__target_datatype" value="biom1"/>
+            <output name="output_biom1">
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/converters/biom.xml]:
```suggestion
            <output name="output" ftype="biom1">
```

File: lib/galaxy/datatypes/converters/biom.xml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                <assert_contents>
+                    <has_text text="Biological Observation Matrix 1.0.0"/>
+                </assert_contents>
+            </output>
+        </test>
+        <test>
+            <param name="input" ftype="biom1" value="input_taxonomy.biom1"/>
+            <param name="__target_datatype" value="biom2"/>
+            <output name="output_biom2" format="biom2" value="input_taxonomy.biom2" compare="sim_size"/>
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/converters/biom.xml]:
```suggestion
            <output name="output" ftype="biom2" value="input_taxonomy.biom2" compare="sim_size"/>
```

File: test/functional/tools/converter_target_datatype.xml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    <!-- test if setting $__target_datatype works for converters -->
+    <command><![CDATA[
+        #if $biom1
+            cp '$biom1' '$output_biom1'
+        #elif $biom2
+            cp '$biom2' '$output_biom2'
+        #else
+            >&2 echo "neither biom1 nor biom2 input received" &&
+            exit 1
+        #end if
```
Comment (nsoranzo) [inline test/functional/tools/converter_target_datatype.xml]:
```suggestion
#if $biom1:
    cp '$biom1' '$output_biom1'
#end if
#if $biom2:
    cp '$biom2' '$output_biom2'
#end if
```

**Comment by bernt-matthias** [comment]
Thanks for the suggestions @nsoranzo .. any preferences about documenting `$__target_datatype`?

File: lib/galaxy/datatypes/converters/biom.xml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    <requirements>
+        <requirement type="package" version="@TOOL_VERSION@">biom-format</requirement>
+    </requirements>
+    <command><![CDATA[
+        biom convert -i '$input' -o '$output'
+        #if $__target_datatype == 'biom2'
+            --to-hdf5
+        #elif $__target_datatype == 'biom1'
+            --to-json
+        #end if
```
Comment (bernt-matthias) [inline lib/galaxy/datatypes/converters/biom.xml:14]:
```suggestion
        #end if
        cp '$ext_config' 'galaxy.json'
```

**Comment by nsoranzo** [comment]
> Thanks for the suggestions @nsoranzo .. any preferences about documenting `$__target_datatype`?

Under https://docs.galaxyproject.org/en/master/dev/schema.html#reserved-variables ?

File: lib/galaxy/datatypes/converters/biom.xml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        cp '$ext_config' 'galaxy.json'
```
Comment (bernt-matthias) [inline lib/galaxy/datatypes/converters/biom.xml]:
```suggestion
        && cp '$ext_config' 'galaxy.json'
```

**Comment by bernt-matthias** [comment]
OK. Guess this fits well for the docs. For consistency I renamed the variable to contain also two trailing underscores.

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 9]
```diff
@@ -3162,6 +3162,7 @@ Name | Description
 ``$__user_id__`` | Email's numeric ID (id column of ``galaxy_user`` table in the database)
 ``$__user_email__`` | User's email address
 ``$__app__`` | The ``galaxy.app.UniverseApplication`` instance, gives access to all other configuration file variables (e.g. $__app__.config.output_size_limit). Should be used as a last resort, may go away in future releases.
+``$__target_datatype__`` | Only available in converters. Contains the target datatype of the conversion
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
``$__target_datatype__`` | Only available in converter tools when run internally by Galaxy. Contains the target datatype of the conversion
```

Maybe also add a small paragraph below suggesting to also add a `__target_datatype__` input param  to be able to run the converter tool manually, and a link to lib/galaxy/datatypes/converters/biom.xml as an example.

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 10]
```diff
@@ -3162,11 +3162,16 @@ Name | Description
 ``$__user_id__`` | Email's numeric ID (id column of ``galaxy_user`` table in the database)
 ``$__user_email__`` | User's email address
 ``$__app__`` | The ``galaxy.app.UniverseApplication`` instance, gives access to all other configuration file variables (e.g. $__app__.config.output_size_limit). Should be used as a last resort, may go away in future releases.
+``$__target_datatype__`` | Only available in converter tools when run internally by Galaxy. Contains the target datatype of the conversion
 
 Additional runtime properties are available as environment variables. Since these
 are not Cheetah variables (the values aren't available until runtime) these should likely
 be escaped with a backslash (``\``) when appearing in ``command`` or ``configfile`` elements.
 
+For internal converter tools using `$__target_datatype__` it is recommended to add a select
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
For internal converter tools using ``$__target_datatype__`` it is recommended to add a select
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 11]
```diff
@@ -3162,11 +3162,16 @@ Name | Description
...
+input parameter with the name `__target_datatype__` in order to make the tool testable, see
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
input parameter with name ``__target_datatype__`` in order to make the tool testable, see
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 12]
```diff
@@ -3162,11 +3162,16 @@ Name | Description
...
+for instance the [biom converter](https://github.com/galaxyproject/galaxy/tree/master/lib/galaxy/datatypes/converters/biom.xml).
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
for instance the [biom converter](https://github.com/galaxyproject/galaxy/blob/master/lib/galaxy/datatypes/converters/biom.xml).
```

File: test/functional/tools/converter_target_datatype.xml [hunk 13]
```diff
@@ -0,0 +1,40 @@
+<tool id="test_CONVERTER_biom" name="Test Convert Biom files" version="0.1">
+    <!-- test if setting $__target_datatype__ works for converters -->
+    <command><![CDATA[
+#if $biom1:
+    cp '$biom1' '$output_biom1'
```
Comment (nsoranzo) [inline test/functional/tools/converter_target_datatype.xml]:
```suggestion
    cp '$biom1' '$output_biom1' &&
```

File: test/functional/tools/converter_target_datatype.xml [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+#end if
+#if $biom2:
+    cp '$biom2' '$output_biom2'
```
Comment (nsoranzo) [inline test/functional/tools/converter_target_datatype.xml]:
```suggestion
    cp '$biom2' '$output_biom2' &&
```

File: test/functional/tools/converter_target_datatype.xml [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+#end if
```
Comment (nsoranzo) [inline test/functional/tools/converter_target_datatype.xml:9]:
```suggestion
#end if
exit 0
```

**Review by nsoranzo — APPROVED** [review]

**Review by mvdbeek — APPROVED** [review]

**Pull request merged**